Submit Homework 4 for Distributed Computing: Deploy a 5-node MariaDB Galera cluster with HAProxy load balancing, automated failover

DmitryFedoroff opened the pull request.

DmitryFedoroff committed:
build: Configure dependencies for container management

#comment Configure Ansible collection requirements for Docker-related modules ensuring community.docker availability

Affected: requirements.yml

DmitryFedoroff committed:
chore: Provide inventory and variable definitions

#comment Provide local connection settings, Python interpreter paths, and WordPress credentials for environment consistency

Affected: hosts.ini, local.yml, all.yml

DmitryFedoroff committed:
feat: Orchestrate site deployment

#comment Orchestrate Ansible roles for Docker, database, and WordPress to enable cohesive setup

Affected: site.yml

DmitryFedoroff committed:
feat: Establish Docker infrastructure

#comment Establish tasks for Docker Engine installation, Compose retrieval, and container orchestration with a dedicated network

Affected: main.yml, docker-compose.yml.j2

DmitryFedoroff committed:
feat: Configure database readiness

#comment Configure waiting mechanism for MySQL and specify credentials to ensure the service is ready before further steps

Affected: main.yml, db-config.yml.j2

DmitryFedoroff committed:
feat: Provision WordPress framework

#comment Provision tasks and templates for custom wp-config with essential database details and environment variables

Affected: main.yml, wp-config.php.j2

DmitryFedoroff committed:
feat: Configure Nginx routing

#comment Configure server block to route requests through PHP-FPM and serve the WordPress files from the correct document root

Affected: nginx.conf

DmitryFedoroff committed:
chore: Add screenshots folder for documentation

#comment Add graphical assets demonstrating deployment state including WordPress installation, admin console, public interface and running containers

Affected: screenshot_01.png, screenshot_02.png, screenshot_03.png, screenshot_04.png

DmitryFedoroff committed:
doc: Document deployment of containerized WordPress infrastructure

#comment Document project architecture, environment requirements, Ansible role structure, Docker service composition and startup procedure with embedded annotated screenshots for verification and reproducibility

Affected: README.md

DmitryFedoroff committed:
feat: Merge HW1 deployment pipeline for containerized WordPress infrastructure

This pull request finalizes Homework 1 for Distributed Computing by delivering a reproducible infrastructure-as-code solution for deploying WordPress using Ansible and Docker.

Highlights:
- Dockerized architecture for MySQL, WordPress (PHP-FPM), and Nginx.
- Ansible roles for Docker installation, service orchestration, and wp-config templating.
- Dynamic inventory and environment variable management.
- Custom Nginx configuration for PHP routing.
- Deployment validated via WordPress UI and CLI state (screenshots included).

DmitryFedoroff committed:
refactor: Update WordPress stack deployment for remote server compatibility and container stability

#comment Refactor WordPress stack deployment:

1. Remove legacy templates wp-config.php.j2 and db-config.yml.j2 to rely on built-in WordPress configuration via environment variables.
2. Configure Docker CE version 28.0.4 and docker-compose-plugin installation via apt with version validation, avoiding insecure curl-based downloads.
3. Relocate nginx.conf into role-specific files directory to follow Ansible structure conventions.
4. Eliminate redundant host_vars/local.yml and consolidate configuration under group_vars/all.yml.
5. Update MySQL, WordPress, and Nginx image tags to fixed stable versions for reproducibility and security.
6. Restrict MySQL service to internal Docker network by removing external port exposure.
7. Implement MySQL healthcheck and enforce startup dependency for WordPress container on database readiness.
8. Ensure persistent volumes are created for both database and application services.
9. Transition Ansible inventory to target remote Ubuntu host via SSH and update playbook to reflect new host group.
10. Maintain idempotency by avoiding shell or command tasks in favor of declarative Ansible modules.
11. Update deployment documentation and visual references via refreshed screenshots.

Affected: group_vars_all_yml, host_vars_local_yml, inventory_hosts_ini, requirements_yml, roles.db_setup.config_files_db_config_yml_j2, roles.db_setup.tasks_main_yml, roles.docker_setup.tasks_main_yml, roles.docker_setup.files_nginx_conf, roles.docker_setup.templates_docker_compose_yml_j2, roles.wordpress_setup.tasks_main_yml, roles.wordpress_setup.templates_wp_config_php_j2, screenshots_screenshot_01_png, screenshots_screenshot_02_png, screenshots_screenshot_03_png, screenshots_screenshot_04_png, site_yml

DmitryFedoroff committed:
refactor: Configure MySQL cluster with GTID replication and legacy data migration

#comment Configure multi-node MySQL cluster with one master and two replicas using GTID-based replication:

1. Replace single MySQL container with multi-node cluster including master and two slaves.
2. Create master.cnf, slave1.cnf and slave2.cnf with GTID, binlog, and replication directives.
3. Adjust docker-compose.yml to define db-master, db-slave1, db-slave2 with configuration volumes and healthchecks.
4. Integrate master/slave MySQL server configs into docker_setup role and copy them to remote host.
5. Detect legacy MySQL volume, run temporary container and dump all data into the new master.
6. Create replication user inside the master node using docker_container_exec.
7. Start replication on both slaves using CHANGE REPLICATION SOURCE TO with GTID-based auto-positioning.
8. Modify db_setup tasks to handle healthcheck, legacy container dump, and slave synchronization.
9. Refactor all.yml to include replication credentials and volumes for cluster migration.

Affected: group_vars/all.yml, docker_setup/tasks/main.yml, docker_setup/templates/docker-compose.yml.j2, docker_setup/files/master.cnf, docker_setup/files/slave1.cnf, docker_setup/files/slave2.cnf, db_setup/tasks/main.yml

DmitryFedoroff committed:
Merge pull request #2 from DmitryFedoroff/feat/hw2-distributed-computing-itmo

feat: Merge HW2 database clustering for MySQL GTID-based replication and legacy migration

DmitryFedoroff committed:
doc: Update README with GTID MySQL cluster architecture and migration flow

#comment Update documentation to reflect improvements in Homework 2 deployment pipeline:

1. Describe transition to distributed MySQL architecture with GTID replication (Master-Slave-Slave).
2. Document legacy volume detection and automated migration strategy via temporary container.
3. Outline Docker Compose network structure, healthchecks, and replication credentials setup.
4. Reflect infrastructure update to remote server with external Nginx port binding and container naming.
5. Clarify WordPress configuration without wp-config.php templating using environment variables.
6. Detail installation instructions including Python venv, Ansible Core, Galaxy collections, and SSH setup.
7. Include updated screenshots of admin UI, Docker containers, and WordPress landing page.
8. Present technical insights into failover, replication mechanics, and compose orchestration model.

Affected: README.md

DmitryFedoroff committed:
Merge pull request #3 from DmitryFedoroff/doc/update-readme-for-gtid-cluster-deployment

doc: Update README with GTID MySQL cluster architecture and migration flow

DmitryFedoroff committed:
refactor: Replace unsafe command executions with structured Docker exec calls

#comment Refactor MySQL replication setup by eliminating all shell-style command invocations:

1. Convert all mysql calls to argv format for safer container execution
2. Replace sh -c with explicit /bin/bash only where piping requires shell
3. Ensure all changes are restricted to docker_container_exec usage
4. Maintain idempotent logic and original orchestration flow

Affected: roles/db_setup/tasks/main.yml

DmitryFedoroff committed:
refactor: Replace unsafe command executions with structured Docker exec calls

#comment Refactor MySQL replication setup by eliminating all shell-style command invocations:

1. Convert all mysql calls to argv format for safer container execution
2. Replace sh -c with explicit /bin/bash only where piping requires shell
3. Ensure all changes are restricted to docker_container_exec usage
4. Maintain idempotent logic and original orchestration flow

Affected: roles/db_setup/tasks/main.yml

DmitryFedoroff committed:
refactor: Transition to declarative MySQL cluster management with community.mysql

#comment Refactor infrastructure code to replace imperative docker_container_exec usage with safe and idempotent community.mysql modules for MySQL user creation, data import, and replication setup. Include dependency on community.mysql, enable PyMySQL, switch to declarative dump/import, and rework docker-compose for external network compatibility.

Affected: requirements.yml, db_setup/tasks/main.yml, docker_setup/tasks/main.yml, docker_setup/templates/docker-compose.yml.j2

DmitryFedoroff committed:
refactor: Refine Docker and MySQL roles for idempotent secure replication

#comment Refactor installation and replication logic to increase idempotency, security and observability:

1. Collect package facts via apt and expose current docker‑ce and compose plugin versions for operators.
2. Verify version compliance against 5:28.0.4 and 2.20.0 thresholds and conditionally add Docker APT repo and GPG key.
3. Separate engine and plugin installation paths using docker_pkg_version and compose_plugin_min_version facts.
4. Skip redundant package upgrades when required versions already present to keep reruns fast and safe.
5. Recreate replication account with REQUIRE SSL and GTID‑safe grants to enforce encrypted channels.
6. Hard reset replica state on slaves and apply CHANGE PRIMARY with primary_ssl for auto‑positioned GTID replication.
7. Insert pause to allow catch‑up and assert IO and SQL thread health with explicit fail‑fast checks.
8. Validate wordpress_db existence on each replica and report Seconds_Behind_Source for troubleshooting.
9. Output informative debug messages to clarify play behaviour during dry‑runs and real executions.

Affected: docker_setup/tasks/main.yml, db_setup/tasks/main.yml

DmitryFedoroff committed:
Merge branch 'main' into feat/hw2-distributed-computing-itmo

DmitryFedoroff committed:
refactor: Refine Docker and MySQL roles for idempotent secure replication

#comment Refactor installation and replication logic to increase idempotency, security and observability:

1. Collect package facts via apt and expose current docker‑ce and compose plugin versions for operators.
2. Verify version compliance against 5:28.0.4 and 2.20.0 thresholds and conditionally add Docker APT repo and GPG key.
3. Separate engine and plugin installation paths using docker_pkg_version and compose_plugin_min_version facts.
4. Skip redundant package upgrades when required versions already present to keep reruns fast and safe.
5. Recreate replication account with REQUIRE SSL and GTID‑safe grants to enforce encrypted channels.
6. Hard reset replica state on slaves and apply CHANGE PRIMARY with primary_ssl for auto‑positioned GTID replication.
7. Insert pause to allow catch‑up and assert IO and SQL thread health with explicit fail‑fast checks.
8. Validate wordpress_db existence on each replica and report Seconds_Behind_Source for troubleshooting.
9. Output informative debug messages to clarify play behaviour during dry‑runs and real executions.

Affected: docker_setup/tasks/main.yml, db_setup/tasks/main.yml

DmitryFedoroff committed:
refactor: Integrate GTID replication and monitoring stack for distributed WordPress cluster

#comment Refactor WordPress deployment and monitoring infrastructure for Homework 3 of Distributed Computing course:

1. Modularize MySQL replication logic into reusable task configure_replica.yml with GTID and health validation.
2. Introduce cAdvisor container for Docker-level performance metrics.
3. Provision Prometheus service with external scrape configuration for cAdvisor.
4. Provision Grafana container with predefined Prometheus datasource.
5. Automate WordPress container initialization with WP-CLI core installation.
6. Refactor docker_setup to ensure idempotent installation of Docker, Compose and SDKs.
7. Consolidate ansible.cfg and inventory.yml to support password-based connection with paramiko.
8. Remove deprecated site.yml and static inventory/hosts.ini
9. Improve playbooks modularity and role separation across playbook1.yml, playbook2.yml, and playbook3.yml

Affected: ansible.cfg, group_vars/all.yml, inventory.yml, playbook1.yml, playbook2.yml, playbook3.yml, requirements.yml, roles/cadvisor/tasks/main.yml, roles/db_setup/tasks/configure_replica.yml, roles/db_setup/tasks/main.yml, roles/docker_setup/tasks/main.yml, roles/grafana/files/datasource.yml, roles/grafana/tasks/main.yml, roles/prometheus/files/prometheus.yml, roles/prometheus/tasks/main.yml, roles/wordpress_setup/tasks/main.yml, .ansible-lint.yaml, .github/workflows/ansible-lint.yml, site.yml, inventory/hosts.ini

DmitryFedoroff committed:
Submit Homework 3 for Distributed Computing: Monitoring-enabled WordPress cluster with GTID-based MySQL replication using Ansible and Docker

- Modularize MySQL replication logic into reusable task configure_replica.yml with GTID and health validation.
- Introduce cAdvisor container for Docker-level performance metrics.
- Provision Prometheus service with external scrape configuration for cAdvisor.
- Provision Grafana container with predefined Prometheus datasource.
- Automate WordPress container initialization with WP-CLI core installation.
- Refactor docker_setup to ensure idempotent installation of Docker, Compose and SDKs.
- Consolidate ansible.cfg and inventory.yml to support password-based connection with paramiko.
- Remove deprecated site.yml and static inventory/hosts.ini
- Improve playbooks modularity and role separation across playbook1.yml, playbook2.yml, and playbook3.yml

DmitryFedoroff committed:
refactor: Consolidate Grafana provisioning and refine MySQL replication control logic

#comment Refactor monitoring and replication subsystems as follows:

1. Configure automated Grafana dashboard provisioning via dashboards.yml and a JSON-based metrics panel layout.
2. Refactor Grafana container mounting logic to include dashboards volume.
3. Improve readability of Grafana tasks by separating dashboard provisioning paths.
4. Handle edge-case failures during MySQL stopreplica stage using safe conditional logic.
5. Extract startreplica execution into dedicated handler for modular reuse.
6. Validate database existence on slaves after GTID replication is established.
7. Refactor docker_setup tasks to unify version check logic and simplify debug output.
8. Update group_vars to include structured Grafana dashboard path variable and correct WordPress admin email.
9. Adjust WP core install logic to support idempotent deployment and clearly interpret exit codes.

Affected: roles/grafana/files/datasource.yml, roles/grafana/files/dashboards.yml, roles/grafana/files/dashboard.json, roles/grafana/tasks/main.yml, roles/db_setup/tasks/configure_replica.yml, roles/db_setup/handlers/main.yml, roles/db_setup/tasks/main.yml, roles/docker_setup/tasks/main.yml, roles/wordpress_setup/tasks/main.yml, group_vars/all.yml, inventory.yml, requirements.yml

DmitryFedoroff committed:
Merge pull request #7 from DmitryFedoroff/feat/hw3-distributed-computing-itmo

refactor: Consolidate Grafana provisioning and refine MySQL replication control logic

DmitryFedoroff committed:
Add .gitignore

DmitryFedoroff committed:
refactor: Delegate Galera cluster deployment to reusable role

#comment Replace inline Galera cluster deployment logic with an import of a reusable Ansible role to improve maintainability and align with role-based design pattern.

Affected: playbook4.yml

DmitryFedoroff committed:
feat: Wait for Galera node readiness during container startup

#comment Start Galera container with required configuration, monitor its health status, update /etc/hosts with its IP, and block execution until MySQL port is accessible.

Affected: wait_container_healthy.yml

DmitryFedoroff committed:
feat: Provision Galera dashboard in Grafana

#comment Copy Galera dashboard JSON to Grafana provisioning directory with correct permissions and trigger Grafana reload to apply updates.

Affected: update_grafana_dashboard.yml

DmitryFedoroff committed:
chore: Restart observability and load balancer containers

#comment Ensure HAProxy, Grafana, and Prometheus containers are restarted via Ansible to reapply configuration or recover from transient issues.

Affected: main.yml

DmitryFedoroff committed:
chore: Standardize quotes in Ansible inventory variable

#comment Replace single quotes with double quotes for ansible_ssh_common_args to maintain YAML consistency and improve editor compatibility.

Affected: inventory.yml

DmitryFedoroff committed:
feat: Provide Grafana dashboard for HAProxy and Galera monitoring

#comment Define Grafana dashboard JSON with panels for Galera node availability, active sessions, and HAProxy server health using Prometheus data source.

Affected: haproxy_stats_dashboard.json

DmitryFedoroff committed:
feat: Provide Grafana dashboard to monitor Prometheus targets

#comment Define JSON dashboard with stat and table panels showing availability of Prometheus exporters using templated queries and threshold mappings.

Affected: prometheus_targets_dashboard.json

DmitryFedoroff committed:
feat: Bootstrap primary Galera node and ensure full cluster health

#comment Launch first Galera container with cluster bootstrap parameters for fresh deployment or migration, and iterate over all nodes to verify container availability and health status.

Affected: manage_galera.yml

DmitryFedoroff committed:
feat: Define centralized configuration for Galera, HAProxy, and monitoring stack

#comment Declare all key variables for database, load balancer, exporters, and dashboards including passwords, image names, ports, and paths to support stack-wide orchestration.

Affected: main.yml

DmitryFedoroff committed:
feat: Orchestrate Galera deployment, migration, monitoring, and cleanup

#comment Coordinate all provisioning tasks including environment preparation, cluster deployment, MySQL migration, monitoring setup, HAProxy deployment, and system cleanup through centralized playbook.

Affected: main.yml

DmitryFedoroff committed:
feat: Prepare environment and detect existing database nodes

#comment Install MariaDB client, create Docker network, render HAProxy config, detect legacy MySQL and Galera nodes, set relevant Ansible facts, and update /etc/hosts for legacy support.

Affected: prepare.yml

DmitryFedoroff committed:
chore: Remove stopped and dangling containers using Docker prune

#comment Clean up unused containers to reduce clutter and free up resources without affecting images, volumes, or networks.

Affected: cleanup_prune.yml

DmitryFedoroff committed:
feat: Configure HAProxy to load balance Galera cluster with health checks

#comment Define HAProxy configuration template with frontend listener, backend roundrobin balancing, MySQL health checks, and stats interface for Galera nodes.

Affected: haproxy.cfg.j2

DmitryFedoroff committed:
feat: Configure Prometheus to scrape metrics from HAProxy and Galera exporters

#comment Define dynamic Prometheus configuration template with scrape targets for cAdvisor, HAProxy exporter, and MariaDB exporters associated with Galera nodes.

Affected: prometheus.yml.j2

DmitryFedoroff committed:
feat: Migrate WordPress database from legacy MySQL to Galera cluster

#comment Dump WordPress database from single-master MySQL instance, remove legacy containers, wait for Galera readiness, import dump into Galera node, and clean up temporary files.

Affected: migrate.yml

DmitryFedoroff committed:
feat: Deploy HAProxy container to load balance Galera cluster

#comment Start or recreate HAProxy container with mounted configuration, exposed ports for MySQL and stats interface, and ensure readiness by waiting for both ports to be accessible.

Affected: deploy_haproxy.yml

DmitryFedoroff committed:
feat: Deploy Prometheus exporters for HAProxy and Galera cluster

#comment Start HAProxy and mysqld exporters as Docker containers, render Prometheus configuration with dynamic targets, and restart Prometheus if the configuration has changed.

Affected: deploy_exporters.yml

DmitryFedoroff committed:
feat: Create MySQL users for monitoring exporters and HAProxy checks

#comment Create Galera users for HAProxy health check and Prometheus mysqld_exporter with appropriate privileges for remote access and metrics collection.

Affected: create_monitoring_users.yml

DmitryFedoroff committed:
chore: Map galera_cluster variables to standard names for backward compatibility

#comment Use conditional set_fact to assign legacy galera_cluster_* variable values to updated variable names when defaults are not explicitly provided.

Affected: compat.yml

DmitryFedoroff committed:
feat: Provide Grafana dashboard for Galera cluster and HAProxy observability

#comment Define JSON dashboard with multiple Prometheus-powered panels covering cluster status, replication metrics, HAProxy statistics, resource usage, and dynamic instance filtering.

Affected: galera_dashboard.json

DmitryFedoroff committed:
refactor: Restart Prometheus container only if configuration has changed

#comment Register config copy task output and conditionally restart Prometheus container to avoid unnecessary restarts and improve idempotency.

Affected: main.yml

DmitryFedoroff committed:
feat: Extend Prometheus configuration to monitor HAProxy and Galera nodes

#comment Add new scrape jobs for HAProxy exporter and MariaDB exporters with adjusted intervals and timeouts to improve observability and coverage of core services.

Affected: prometheus.yml

DmitryFedoroff committed:
refactor: Improve cAdvisor container behavior with hostname and compatibility mode

#comment Set explicit container hostname and enable compatibility mode to ensure stable deployment and consistent network identification for monitoring purposes.

Affected: main.yml

DmitryFedoroff committed:
feat: Extend global configuration with Galera cluster and monitoring variables

#comment Define root password mapping, Galera node hosts, HAProxy exporter ports, cluster image and SST method, exporter credentials, and unify root password reference for consistent usage.

Affected: all.yml

DmitryFedoroff committed:
doc: Update README with latest configuration and usage details

#comment Revise README file to reflect recent changes in project setup, deployment steps, and environment variables.

Affected: README.md

SemenMartynov commented:
Что-то у меня на 1-ом плейбуке уже рухнуло
```
TASK [wordpress_setup : Install WordPress core] ******************************************************************************************************************************************************
fatal: [carrier]: FAILED! => {"changed": false, "rc": 1, "stderr": "Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or that contact with the database server at `db-master` could not be established. This could mean your host’s database server is down.", "stderr_lines": ["Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or that contact with the database server at `db-master` could not be established. This could mean your host’s database server is down."], "stdout": "", "stdout_lines": []}
```

При повторном запуске
```
TASK [wordpress_setup : Install WP-CLI in container (idempotent)] ************************************************************************************************************************************
task path: /workspaces/IFMO_DistributedComputing_for_DevOps/roles/wordpress_setup/tasks/main.yml:14
fatal: [carrier]: FAILED! => {"changed": false, "rc": 1, "stderr": "chmod: cannot access 'wp-cli.phar': No such file or directory", "stderr_lines": ["chmod: cannot access 'wp-cli.phar': No such file or directory"], "stdout": "/usr/local/bin/wp", "stdout_lines": ["/usr/local/bin/wp"]}
```

DmitryFedoroff commented:
> Что-то у меня на 1-ом плейбуке уже рухнуло
> 
> ```
> TASK [wordpress_setup : Install WordPress core] ******************************************************************************************************************************************************
> fatal: [carrier]: FAILED! => {"changed": false, "rc": 1, "stderr": "Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or that contact with the database server at `db-master` could not be established. This could mean your host’s database server is down.", "stderr_lines": ["Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or that contact with the database server at `db-master` could not be established. This could mean your host’s database server is down."], "stdout": "", "stdout_lines": []}
> ```
> 
> При повторном запуске
> 
> ```
> TASK [wordpress_setup : Install WP-CLI in container (idempotent)] ************************************************************************************************************************************
> task path: /workspaces/IFMO_DistributedComputing_for_DevOps/roles/wordpress_setup/tasks/main.yml:14
> fatal: [carrier]: FAILED! => {"changed": false, "rc": 1, "stderr": "chmod: cannot access 'wp-cli.phar': No such file or directory", "stderr_lines": ["chmod: cannot access 'wp-cli.phar': No such file or directory"], "stdout": "/usr/local/bin/wp", "stdout_lines": ["/usr/local/bin/wp"]}
> ```

Сейчас постараюсь найти ошибку, но сделать это будет думаю непросто, потому что я не сталкиваюсь с этой проблой. Однако это обычная практика

DmitryFedoroff committed:
fix: Correct WordPress database connection and provision monitoring users on all nodes

#comment Fix critical WordPress database connection error by using correct db_host variable instead of hardcoded db-master, ensuring WordPress connects to mysql_master container. Refactor monitoring user creation to provision users on all Galera nodes for proper HAProxy health checks during roundrobin balancing:

1. Fix docker-compose.yml.j2 to use {{ db_host }} variable for WORDPRESS_DB_HOST instead of hardcoded db-master, resolving Error establishing a database connection issue
2. Update wordpress_setup tasks to wait for {{ db_host }} container instead of hardcoded mysql_master, ensuring proper synchronization with actual container name
3. Correct WP-CLI installation command to handle idempotency properly by checking if wp command exists before attempting download
4. Modify create_monitoring_users.yml to loop through all galera_nodes when creating HAProxy cluster-check user, preventing health check failures on non-primary nodes
5. Update create_monitoring_users.yml to provision exporter user on all Galera nodes with required privileges for metrics collection
6. Pin mariadb_exporter_image to prom/mysqld-exporter:v0.14.0 for version stability and predictable behavior
7. Standardize mysqld-exporter references in prometheus.yml.j2 to use consistent mariadb_exporter naming
8. Maintain datasource UID consistency in prometheus_targets_dashboard.json for proper Grafana integration
9. Ensure WordPress can connect to database on first playbook run and all subsequent runs without connection errors

Affected: docker-compose.yml.j2, roles/galera_cluster/defaults/main.yml, roles/wordpress_setup/tasks/main.yml, create_monitoring_users.yml, prometheus_targets_dashboard.json, prometheus.yml.j2

DmitryFedoroff commented:
Панель Grafana "Prometheus Targets - Advanced Monitoring" на http://83.220.168.126:3000/d/prometheus-targets-advanced/prometheus-targets-advanced-monitoring c общим состоянием 7-ми таргетов, включая 5 mysql_galera, со 100 % доступностью.
![Screenshot_from_2025-05-25-19-28-23_prometheus_targets-advanced_monitoring_5_ok](https://github.com/user-attachments/assets/1b5baccd-e154-46a2-9f67-d3ff31aaf2c3)

HAProxy статистика на http://83.220.168.126:8404/stats c функционирующим фронтендом и бэкенд с 5-ю узлами в состоянии UP.
![Screenshot_from-2025-05-25_19-30-48_ha_stats_5_ok_before](https://github.com/user-attachments/assets/8c93485c-b814-493a-984e-ddb3be6700a9)

Здесь состояние после имитации сбоя 2-х узлов (galera_node_4 и galera_node_5) переведены в DOWN, а остальные 3 продолжают обслуживать трафик.
![Screenshot_from_2025-05-25_19-37-15_ha_stats_3_ok_kill](https://github.com/user-attachments/assets/933cda4d-ee0d-44d7-bf49-2477b5615065)

После проверки отказоустойчивости статистика HAProxy показывает все 5 узлов Galera в статусе UP. Данные снова начали поступать на galera_node_4 и galera_node_5.
![Screenshot_from_2025-05-25_19-49-21_ha_stats_5_ok_after](https://github.com/user-attachments/assets/0a24cf1d-0356-407c-aa20-5c151447b38d)

Интерфейс Prometheus на http://83.220.168.126:9091/targets также подтверждает состояние UP для всех таргетов cadvisor, haproxy и mysql_galera.
![Screenshot_from_2025-05-25_19-19-11_targets](https://github.com/user-attachments/assets/8b45ef39-aa83-45e8-a981-8f2d44a35c70)

DmitryFedoroff commented:
> Что-то у меня на 1-ом плейбуке уже рухнуло
> 
> ```
> TASK [wordpress_setup : Install WordPress core] ******************************************************************************************************************************************************
> fatal: [carrier]: FAILED! => {"changed": false, "rc": 1, "stderr": "Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or that contact with the database server at `db-master` could not be established. This could mean your host’s database server is down.", "stderr_lines": ["Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or that contact with the database server at `db-master` could not be established. This could mean your host’s database server is down."], "stdout": "", "stdout_lines": []}
> ```
> 
> При повторном запуске
> 
> ```
> TASK [wordpress_setup : Install WP-CLI in container (idempotent)] ************************************************************************************************************************************
> task path: /workspaces/IFMO_DistributedComputing_for_DevOps/roles/wordpress_setup/tasks/main.yml:14
> fatal: [carrier]: FAILED! => {"changed": false, "rc": 1, "stderr": "chmod: cannot access 'wp-cli.phar': No such file or directory", "stderr_lines": ["chmod: cannot access 'wp-cli.phar': No such file or directory"], "stdout": "/usr/local/bin/wp", "stdout_lines": ["/usr/local/bin/wp"]}
> ```

Прошу вас проверить 4-е домашнее задание.
Правки внес и для демонстрации прикладываю скриншоты.

SemenMartynov commented:
Теперь упал 4ый плейбук
```
TASK [galera_cluster : Import dump into first Galera node] *******************************************************************************************************************************************
task path: /workspaces/IFMO_DistributedComputing_for_DevOps/roles/galera_cluster/tasks/migrate.yml:32
fatal: [carrier]: FAILED! => {"changed": false, "msg": "unable to find . Exception message: (1130, \"172.18.0.1' is not allowed to connect to this MariaDB server\")"}
```

При повторном запуске
```
TASK [galera_cluster : Create HAProxy cluster-check user (no password) on all Galera nodes] **********************************************************************************************************
task path: /workspaces/IFMO_DistributedComputing_for_DevOps/roles/galera_cluster/tasks/create_monitoring_users.yml:2
failed: [carrier] (item={'name': 'galera_node_1', 'id': 1}) => {"ansible_loop_var": "item", "changed": false, "item": {"id": 1, "name": "galera_node_1"}, "msg": "unable to connect to database, check login_user and login_password are correct or /root/.my.cnf has the credentials. Exception message: (1130, \"172.18.0.1' is not allowed to connect to this MariaDB server\")"}
failed: [carrier] (item={'name': 'galera_node_2', 'id': 2}) => {"ansible_loop_var": "item", "changed": false, "item": {"id": 2, "name": "galera_node_2"}, "msg": "unable to connect to database, check login_user and login_password are correct or /root/.my.cnf has the credentials. Exception message: (1130, \"172.18.0.1' is not allowed to connect to this MariaDB server\")"}
failed: [carrier] (item={'name': 'galera_node_3', 'id': 3}) => {"ansible_loop_var": "item", "changed": false, "item": {"id": 3, "name": "galera_node_3"}, "msg": "unable to connect to database, check login_user and login_password are correct or /root/.my.cnf has the credentials. Exception message: (1130, \"172.18.0.1' is not allowed to connect to this MariaDB server\")"}
failed: [carrier] (item={'name': 'galera_node_4', 'id': 4}) => {"ansible_loop_var": "item", "changed": false, "item": {"id": 4, "name": "galera_node_4"}, "msg": "unable to connect to database, check login_user and login_password are correct or /root/.my.cnf has the credentials. Exception message: (1130, \"172.18.0.1' is not allowed to connect to this MariaDB server\")"}
failed: [carrier] (item={'name': 'galera_node_5', 'id': 5}) => {"ansible_loop_var": "item", "changed": false, "item": {"id": 5, "name": "galera_node_5"}, "msg": "unable to connect to database, check login_user and login_password are correct or /root/.my.cnf has the credentials. Exception message: (1130, \"172.18.0.1' is not allowed to connect to this MariaDB server\")"}
```

На всякий случай, проверил с консольки, что подключение не разрешено...
Так что увы, на скриншоты полагаться не получится)

DmitryFedoroff committed:
fix: Allow root user connections from any host in Galera cluster

#comment Configure MARIADB_ROOT_HOST environment variable to '%' in all Galera nodes to resolve connection rejection error (1130) when Ansible attempts to connect from Docker host IP address.

Affected: manage_galera.yml, wait_container_healthy.yml

DmitryFedoroff commented:
Я вас понимаю, но я все же буду добавлять скриншоты, чтобы это не выглядело так, что я присылаю вам нерабочее домашнее задание.

![Screenshot from 2025-05-26 02-45-41](https://github.com/user-attachments/assets/b282273b-4e51-496e-94e2-cc63360e8c14)

![Screenshot from 2025-05-26 02-45-53](https://github.com/user-attachments/assets/14e0b677-f093-4b2d-a6d2-1bc0cacd0e95)

![Screenshot from 2025-05-26 02-46-01](https://github.com/user-attachments/assets/8e4021f1-52dc-415d-afe4-e215f69d8f4c)

![Screenshot from 2025-05-26 02-46-09](https://github.com/user-attachments/assets/a786c608-5248-46bf-8ec1-4e94ba759abc)

DmitryFedoroff commented:
Прошу вас проверить 4-е домашнее задание.

SemenMartynov commented:
А что такое `MARIADB_ROOT_HOST`?
В доках про такую переменную ничего не сказано -- https://hub.docker.com/r/bitnami/mariadb-galera

Так что всё по прежнему:
```
TASK [galera_cluster : Import dump into first Galera node] *******************************************************************************************************************************************
task path: /workspaces/IFMO_DistributedComputing_for_DevOps/roles/galera_cluster/tasks/migrate.yml:32
fatal: [carrier]: FAILED! => changed=false 
  msg: 'unable to find . Exception message: (1130, "172.18.0.1'' is not allowed to connect to this MariaDB server")'
```

DmitryFedoroff committed:
refactor: Update host IP in Ansible inventory

#comment Replace outdated target IP address in inventory.yml with new operational address to reflect the current remote deployment host. Maintain existing Ansible connection and privilege escalation settings.

Affected: inventory.yml

DmitryFedoroff committed:
refactor: Harden Docker setup and improve APT initialization

#comment Create missing APT directories, update cache explicitly before Docker installation, and ensure user can interact with Docker without sudo by adding to docker group.

Affected: main.yml

DmitryFedoroff committed:
feat: Display Prometheus targets via new Grafana dashboard

#comment Copy Prometheus targets dashboard JSON to Grafana dashboards directory to visualize monitored endpoints alongside existing container metrics.

Affected: main.yml

DmitryFedoroff committed:
refactor: Pin MariaDB Galera Docker image to version 11.0

#comment Replace floating latest tag with explicit 11.0 version for Galera image to ensure consistent and reproducible deployments across environments.

Affected: main.yml

DmitryFedoroff committed:
refactor: Strengthen Galera bootstrap, health checks and monitoring integration

#comment Replace deprecated mysql modules with container-level exec commands, implement health checks and wait loops for stability, grant root access dynamically, and remove outdated HAProxy Grafana dashboard.

Affected: main.yml, wait_container_healthy.yml, manage_galera.yml, create_monitoring_users.yml, grafana/files/haproxy_stats_dashboard.json

DmitryFedoroff commented:
Прошу вас проверить 4-ю лабораторную работу.
Взял специально чистую новую виртуальную машину и намеренно послабее предыдущей для чистоты эксперимента, поэтому пришлось поправить ряд файлов. Удалось всё запустить, правда занимает это существенно больше времени. Текущее состояние на скриншотах:

![Screenshot from 2025-06-01 00-45-42](https://github.com/user-attachments/assets/45b94558-f17d-4e76-a713-fd346e60db3d)

![Screenshot from 2025-06-01 00-46-02](https://github.com/user-attachments/assets/593d3091-9bdd-42e2-98f9-6fdff20b1bf6)

![Screenshot from 2025-06-01 00-44-23](https://github.com/user-attachments/assets/c22ad9be-5fd1-40e3-8394-be1b02d8d8e2)

![Screenshot from 2025-06-01 00-45-21](https://github.com/user-attachments/assets/555169a8-9678-47ca-8f5b-bb1fbf791e63)

DmitryFedoroff commented:
Прошу вас проверить 4-е домашнее задание.

SemenMartynov reviewed (approved):
У меня это всё ещё дольше отрабатывает))
Ещё и тесты...
